source/agora/consensus/data/UtxoSet.d

```diff
@@ -0,0 +1,292 @@
+/*******************************************************************************
+
+    Contains an SQLite-backed UTXO transaction set class
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.consensus.data.UTXOSet;
+
+import agora.common.Deserializer;
+import agora.common.Hash;
+import agora.common.Serializer;
+import agora.common.Set;
+import agora.consensus.data.Transaction;
+
+import d2sqlite3.database;
+
+import vibe.core.log;
+
+import std.file;
+
+/// ditto
+public class UTXOSet
+{
+    /// Utxo cache backed by a database
+    private UTXODB utxo_db;
+
+    /// Keeps track of spent outputs during the validation of a Tx / Block
+    private Set!Hash used_utxos;
+
+
+    /***************************************************************************
+
+        Constructor
+
+        Params:
+            utxo_db_path = path to the UTXO database
+
+    ***************************************************************************/
+
+    public this (in string utxo_db_path)
+    {
+        this.utxo_db = new UTXODB(utxo_db_path);
+    }
+
+    /***************************************************************************
+
+        Shut down the utxo store
+
+    ***************************************************************************/
+
+    public void shutdown ()
+    {
+        this.utxo_db.shutdown();
+    }
+
+    /***************************************************************************
+
+        Add all of a transaction's outputs to the Utxo set,
+        and remove the spent outputs in the transaction from the set.
+
+        Params:
+            tx = the transaction
+
+    ***************************************************************************/
+
+    public void updateUtxoCache (const ref Transaction tx) nothrow @safe
+    {
+        foreach (input; tx.inputs)
```

Comment:
`ref` ?

Reply:
For optimization? I could do `const ref` I guess.